Update Notice design to follow Muriel design guidelines

shiki opened the pull request:
Closes #11033.

This changes the design of Notices to follow the latest Muriel design guidelines. This was initiated to complement the planned changes for Offline error alerts to use Notices (#11035).

The changes primarily affect the default `NormalNoticeStyle`. The `QuickStartNoticeStyle` is left unchanged. Though that is already using a dark background.

## ~~Companion Pull Request~~

~~There are changes required in WordPress-iOS-Shared. Please see the PR wordpress-mobile/WordPress-iOS-Shared#186.~~

## Screenshots

Here are some before and after screenshots.

| Before | After |
|--------|-------|
|![2019-02-22 07 45 11](https://user-images.githubusercontent.com/198826/53252333-e1484d00-367b-11e9-837d-51531197cc0d.png)| ![2019-02-25 07 22 17](https://user-images.githubusercontent.com/198826/53343768-5b701000-38ce-11e9-9298-e7f8fd6bc913.png)|
| ![2019-02-22 07 44 09](https://user-images.githubusercontent.com/198826/53252439-1eacda80-367c-11e9-819c-b24c696bd01a.png) |![2019-02-25 07 22 45](https://user-images.githubusercontent.com/198826/53343831-835f7380-38ce-11e9-922d-a3bb664cb494.png) |
| ![2019-02-25 07 41 12](https://user-images.githubusercontent.com/198826/53344851-e3571980-38d0-11e9-846e-211a8f38ed56.png) | ![2019-02-25 07 36 04](https://user-images.githubusercontent.com/198826/53344880-ee11ae80-38d0-11e9-83b9-4bd8afd69424.png) | 

## Testing

You can test this by going over the current usages of Notices like:

- Uploading a media from the Media Library
- Unfollowing a site in Reader → Followed Sites → Manage

You can also try out what we're planning to do with the Offline error notices by:

1. Applying this [diff](https://gist.github.com/shiki/fc81257bf927f7f82b8aba1f725c18d2).
2. Go to Reader
3. Disconnect the device from the Internet

### Possible Regressions

I tested the Quick Start and have verified that nothing changed. Still, I could have missed some things because I am not very familiar with it.

shiki commented:
The tests are failing on `WPAnalyticsTrackerAutomatticTracksTests` which will be fixed once #11060 and #11112 are merged.

shiki commented:
Discussed this with Eric and we'll just add the `WPAnalyticsTrackerAutomatticTracks` event names in the branches. All tests are passing now.

yaelirub commented on WordPress/Classes/ViewRelated/System/Notices/NoticeView.swift:
```diff
@@ -258,16 +258,17 @@ class NoticeView: UIView {
     }
 
     private enum Appearance {
-        static let actionBackgroundColor = UIColor.white.withAlphaComponent(0.5)
```
What was the purpose of this and why was it removed?

shiki replied:
It was a constant and was only used in line 152: 

```swift
actionBackgroundView.backgroundColor = Appearance.actionBackgroundColor
```

But since the `backgroundColor` now uses `notice.style.backgroundColor`, it's no longer needed.

megsfulton commented:
This is looking good, a few small things:

1. For the single line snackbar, let's make the min height on it 48. It's looking a smidge taller than that now. 

2. @iamthomasbishop pointed out that the current color combo isn't passing accessibility contrast standards. Can we update the background color to be Gray 800 instead? That will pass with the medium blue text link. 

3. Let's swap the text weight styling so that the message ("Media uploaded") is regular, and the action ("Write Post") is Medium. That should also help with the legibility of button text.

shiki commented:
@megsfulton 

> 3\. Let's swap the text weight styling so that the message ("Media uploaded") is regular, and the action ("Write Post") is Medium. That should also help with the legibility of button text.

It looks like the message being Medium is because there are 2 types of labels implemented in the current Notice: 

- `title`: Medium
- `message`: Regular

Some Notices use both and they look like this: 

<img src="https://user-images.githubusercontent.com/198826/53270207-55004f00-36a8-11e9-8cb5-cf9380636667.png" width="320">

_(That misplacement is not new)_

What do you suggest for instances like this? What should the text weights be?

nheagy commented on WordPress/Classes/ViewRelated/System/Notices/NoticeView.swift:
```diff
@@ -258,16 +258,17 @@ class NoticeView: UIView {
         static let shadowColor: UIColor = .black
         static let shadowOpacity: Float = 0.25
-        static let shadowRadius: CGFloat = 8.0
+        /// The blur radius of the shadow.
```
I think `shadowRadius` is descriptive enough to not need the comment.

nheagy commented on WordPress/Classes/ViewRelated/System/Notices/NoticeView.swift:
```diff
@@ -258,16 +258,17 @@ class NoticeView: UIView {
     private enum Metrics {
-        static let cornerRadius: CGFloat = 13.0
+        /// The corner radius of the background (colored) and its shadow.
+        static let cornerRadius: CGFloat = 4.0
```
I don't think we need this comment either

shiki replied:
It took me a while to scan the code to figure out how this is used. But yeah, I guess you're right. It's unnecessary noise since it's in the context of `NoticeView`.

nheagy commented on WordPress/Classes/ViewRelated/System/Notices/NoticeStyle.swift:
```diff
@@ -29,11 +29,11 @@ public struct NormalNoticeStyle: NoticeStyle {
-    public let backgroundColor: UIColor = .clear
+    public let titleColor: UIColor = .white
+    public let messageColor: UIColor = .white
+    public let backgroundColor: UIColor = WPStyleGuide.grey700()
```
I totally appreciate that this is a bit awkward because we're using a color from the new palette before the app has the new palette fully implemented. After it's implemented, I think `WPStyleGuide`'s color API will either look a lot different or be replaced entirely.

With that in mind, my preference would be to _not_ change `WordPress-iOS-Shared` and instead make `grey 700` in an extension to `UIColor` just for this file (such as [was done in QuickStartListTitleCell](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/Classes/ViewRelated/Blog/QuickStartListTitleCell.swift#L68)).

I'm just going to leave that as a comment so you can decide to agree or ignore me, as you prefer 😆

shiki replied:
@nheagy Excellent! I understand. This was awkward and I wasn't sure about modifying `WordPress-iOS-Shared`. Since we're already using `UIColor` extensions anyway, I'll do it that way so there's no need to update the pod. Thank you!

megsfulton commented:
>What do you suggest for instances like this? What should the text weights be?

I had forgotten about this case. Instead of what I originally recommended, let's instead make the blue text medium and leave the title/message text as currently is.

shiki commented:
@yaelirub @megsfulton @nheagy Ready for another look. 🙂 I updated the screenshots.

- Changed background color to Gray 800
- Changed action button font to use `medium` weight
- Decreased the vertical margins to make the Notice height closer to 48. It's currently at 49 and it seems dependent on the message's font size. Is this okay @megsfulton?

loremattei commented:
I'm moving this to 12.0 since 11.9 has been cut. If you want it in 11.9, please feel free to target the release branch and ping me to build a new beta.

yaelirub commented on WordPress/Classes/ViewRelated/System/Notices/NoticeStyle.swift:
```diff
@@ -54,3 +54,10 @@ public struct QuickStartNoticeStyle: NoticeStyle {
+private extension UIColor {
+    /// The Gray 800 color (#2b2d2f) from the new color palette.
+    ///
+    /// This is here temporarily until we have the new color palette fully implemented.
```
Maybe add a TODO here?

yaelirub commented on WordPress/Classes/ViewRelated/System/Notices/NoticeView.swift:
```diff
@@ -258,16 +258,15 @@ class NoticeView: UIView {
 
     private enum Metrics {
-        static let cornerRadius: CGFloat = 13.0
+        static let cornerRadius: CGFloat = 4.0
```
I saw @nheagy comment on the corner radius not looking iOS like and I agree. I know that Muriel says 4.0 but thought we agreed to find a middle ground. Are we gonna keep it as 4.0 for now?

shiki replied:
@megsfulton said that we should keep it at 4.0 for now.

megsfulton commented:
Looks good!

designsimply commented:
Tested and confirmed I can see the new notices with dark backgrounds and light readable text in WP Internal 12.0.0.20190311 on iPhone 6S iOS 12.1.4 and wanted to include a few screenshots for reference:

* Scheduled post
* Posts list offline
* Post publish offline

<img src="https://user-images.githubusercontent.com/1119271/54569046-215ecf00-499f-11e9-86af-1743b35b7889.PNG" width="200" alt="IMG_2087"> <img src="https://user-images.githubusercontent.com/1119271/54569173-77337700-499f-11e9-968c-9257911d0ed3.PNG" width="200" alt="IMG_2113"> <img src="https://user-images.githubusercontent.com/1119271/54569382-4bfd5780-49a0-11e9-92b0-d28e51c45961.PNG" width="200" alt="IMG_2116">
